src/Transports.AspNetCore/WebSockets/IWebSocketConnection.cs

```diff
@@ -0,0 +1,50 @@
+using GraphQL.Transport;
+using Microsoft.AspNetCore.Http;
+
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// Represents a WebSocket connection, dispatching received messages over
+/// the connection to the specified <see cref="IOperationMessageProcessor"/>,
+/// and sending requested messages out the connection when requested.
+/// <br/><br/>
+/// <see cref="ExecuteAsync(IOperationMessageProcessor)"/> may be only called once.
+/// All members must be thread-safe.
+/// </summary>
+public interface IWebSocketConnection
+{
+    /// <summary>
+    /// Listens to incoming messages over the WebSocket connection,
+    /// dispatching the messages to the specified <paramref name="operationMessageProcessor"/>.
+    /// Returns or throws <see cref="OperationCanceledException"/> when the WebSocket connection is closed.
+    /// </summary>
+    Task ExecuteAsync(IOperationMessageProcessor operationMessageProcessor);
+
+    /// <summary>
+    /// Sends a message.
+    /// </summary>
+    Task SendMessageAsync(OperationMessage message);
+
+    /// <summary>
+    /// Closes the WebSocket connection.
+    /// </summary>
+    Task CloseConnectionAsync();
+
+    /// <summary>
+    /// Closes the WebSocket connection with the specified error information.
+    /// </summary>
+    Task CloseConnectionAsync(int eventId, string? description);
+
+    /// <summary>
+    /// Returns the last UTC time that a message was sent.
+    /// </summary>
+    DateTime LastMessageSentAt { get; }
+
+    /// <inheritdoc cref="HttpContext.RequestAborted"/>
+    CancellationToken RequestAborted { get; }
```

Comment:
Why do you need it? `HttpContext` is already here.

Reply:
Or is that `HttpContext` Initial one where connection was created?

Reply:
Because shutting down the host typically waits for active HTTP connections to complete, up to 30 seconds.  GraphQL subscriptions would prevent prompt shutdown, typically preventing shutdown for the full 30 seconds, but by monitoring `IHostApplicationLifetime`, we can shut down WebSocket connections promptly when the host is shutting down.

Note that while it is possible to attempt a graceful shutdown of the WebSocket connection upon a host shutdown notification, I did not attempt it.  Rather, if the host is shutting down, all subscriptions are immediately disconnected, and the HTTP connection is immediately shut down.  This is accomplished by creating a linked cancellation token source and using it throughout the WebSocket pipeline rather than `HttpContext.RequestAborted`.

See also:
- https://github.com/ChilliCream/hotchocolate/pull/4775

Reply:
> Or is that `HttpContext` initial one where connection was created?

Yes... there is only a single `HttpContext` (and a single `WebSocket` instance) for the entire lifetime of a WebSocket connection over HTTP.

Reply:
Thanks for explanations. So actually `RequestAborted` does not equal to `HttpContext.RequestAborted`. It may be a some linked token, right?. If so maybe to give some other name to this property to not interfere with `HttpContext.RequestAborted` ?

Reply:
I’m open to suggestions. It’s not like the name is “ConnectionClosed” - the application shutdown is aborting the request so the terminology is still correct - but we could change it if you think we should.  (I’d leave it.)